cmd/kube-controller-manager/app/core.go

```diff
@@ -585,50 +573,58 @@ func processCIDRs(cidrsList string) ([]*net.IPNet, bool, error) {
 	return cidrs, dualstack, nil
 }
 
-// setNodeCIDRMaskSizes returns the IPv4 and IPv6 node cidr mask sizes.
-// If --node-cidr-mask-size not set, then it will return default IPv4 and IPv6 cidr mask sizes.
-func setNodeCIDRMaskSizes(cfg nodeipamconfig.NodeIPAMControllerConfiguration) (int, int, error) {
+// getNodeCIDRMaskSizes returns the node cidr mask sizes slice based on the cluster cidr slice
+func getNodeCIDRMaskSizes(clusterCIDRs []*net.IPNet, cfg nodeipamconfig.NodeIPAMControllerConfiguration) ([]int, error) {
+	nodeMaskCIDRs := make([]int, len(clusterCIDRs))
 	ipv4Mask, ipv6Mask := defaultNodeMaskCIDRIPv4, defaultNodeMaskCIDRIPv6
-	// NodeCIDRMaskSizeIPv4 and NodeCIDRMaskSizeIPv6 can be used only for dual-stack clusters
-	if cfg.NodeCIDRMaskSizeIPv4 != 0 || cfg.NodeCIDRMaskSizeIPv6 != 0 {
-		return ipv4Mask, ipv6Mask, errors.New("usage of --node-cidr-mask-size-ipv4 and --node-cidr-mask-size-ipv6 are not allowed with non dual-stack clusters")
-	}
-	if cfg.NodeCIDRMaskSize != 0 {
-		ipv4Mask = int(cfg.NodeCIDRMaskSize)
-		ipv6Mask = int(cfg.NodeCIDRMaskSize)
-	}
-	return ipv4Mask, ipv6Mask, nil
-}
 
-// setNodeCIDRMaskSizesDualStack returns the IPv4 and IPv6 node cidr mask sizes to the value provided
-// for --node-cidr-mask-size-ipv4 and --node-cidr-mask-size-ipv6 respectively. If value not provided,
-// then it will return default IPv4 and IPv6 cidr mask sizes.
-func setNodeCIDRMaskSizesDualStack(cfg nodeipamconfig.NodeIPAMControllerConfiguration) (int, int, error) {
-	ipv4Mask, ipv6Mask := defaultNodeMaskCIDRIPv4, defaultNodeMaskCIDRIPv6
-	// NodeCIDRMaskSize can be used only for single stack clusters
-	if cfg.NodeCIDRMaskSize != 0 {
-		return ipv4Mask, ipv6Mask, errors.New("usage of --node-cidr-mask-size is not allowed with dual-stack clusters")
-	}
-	if cfg.NodeCIDRMaskSizeIPv4 != 0 {
-		ipv4Mask = int(cfg.NodeCIDRMaskSizeIPv4)
-	}
-	if cfg.NodeCIDRMaskSizeIPv6 != 0 {
-		ipv6Mask = int(cfg.NodeCIDRMaskSizeIPv6)
+	switch len(clusterCIDRs) {
+	case 1:
+		if netutils.IsIPv6CIDR(clusterCIDRs[0]) {
+			if cfg.NodeCIDRMaskSizeIPv4 != 0 {
+				return nodeMaskCIDRs, errors.New("specifying --node-cidr-mask-size-ipv4 is not allowed with IPv6 single-stack clusters")
+			}
+			if cfg.NodeCIDRMaskSize != 0 && cfg.NodeCIDRMaskSizeIPv6 != 0 && cfg.NodeCIDRMaskSize != cfg.NodeCIDRMaskSizeIPv6 {
+				return nodeMaskCIDRs, errors.New("specifying both --node-cidr-mask-size and --node-cidr-mask-size-ipv6 with different values is not allowed with IPv6 single-stack clusters")
```

Comment:
There's no real reason to allow specifying both of them with the same value. Just make it an error if they're both set, regardless of whether they're set to the same or different values